pkg/virt-controller/watch/ovm.go

```diff
@@ -0,0 +1,184 @@
+package watch
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	runtime "k8s.io/apimachinery/pkg/util/runtime"
+	wait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	workqueue "k8s.io/client-go/util/workqueue"
+
+	kubeapiv1 "kubevirt.io/kubevirt/pkg/api/v1"
+	kubecli "kubevirt.io/kubevirt/pkg/kubecli"
+)
+
+const controllerName = "offlinevm-controller"
+
+// OVMController watches the offline virtual machines and keeps them in
+// sync with the Virtual Machine. The Virtual Machine is created by this
+// controller if needed
+type OVMController struct {
+	// client which connects to cluster with all kubevirt added
+	client kubecli.KubevirtClient
+
+	ovmInformer cache.SharedIndexInformer
+	vmInformer  cache.SharedIndexInformer
+
+	// workque is used to buffer and limit work ammount. We do not want to be
+	// too fast
+	workqueue workqueue.RateLimitingInterface
+
+	// recorder and broadcaster handle events from and to cluster
+	eventRecorder record.EventRecorder
+}
+
+// NewOVMController creates instance of OVM controller and setups the required stuff
+func NewOVMController(
+	clientset kubecli.KubevirtClient,
+	recorder record.EventRecorder,
+	ovmInformer cache.SharedIndexInformer,
+	vmInformer cache.SharedIndexInformer,
+) *OVMController {
+
+	controller := &OVMController{
+		client:        clientset,
+		ovmInformer:   ovmInformer,
+		vmInformer:    vmInformer,
+		workqueue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ovms"),
+		eventRecorder: recorder,
+	}
+
+	// Register the listener for OVM kind events in the cluster
+	ovmInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
```

Comment:
You will need pretty much a 1:1 copy for the callbacks, like the look lie in the VirtualMachineReplicaSet controller. That includes reference management and the expectation manager.